homeassistant/components/rainbird/config_flow.py

```diff
@@ -0,0 +1,164 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): selector.TextSelector(),
+        vol.Required(CONF_PASSWORD): selector.TextSelector(
+            selector.TextSelectorConfig(type=selector.TextSelectorType.PASSWORD)
+        ),
+    }
+)
+
+
+class ConfigFlowError(Exception):
+    """Error raised during a config flow."""
+
+    def __init__(self, message: str, error_code: str) -> None:
+        """Initialize ConfigFlowError."""
+        super().__init__(message)
+        self.error_code = error_code
+
+
+class RainbirdConfigFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Rain Bird."""
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: ConfigEntry,
+    ) -> RainBirdOptionsFlowHandler:
+        """Define the config flow to handle options."""
+        return RainBirdOptionsFlowHandler(config_entry)
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Configure the Rain Bird device."""
+        error_code: str | None = None
+        if user_input:
+            try:
+                serial_number = await self._test_connection(
+                    user_input[CONF_HOST], user_input[CONF_PASSWORD]
+                )
+            except ConfigFlowError as err:
+                _LOGGER.error("Error during config flow: %s", err)
+                error_code = err.error_code
+            else:
+                return await self.async_finish(serial_number, user_input)
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=DATA_SCHEMA,
+            errors={"base": error_code} if error_code else None,
+        )
+
+    async def _test_connection(self, host: str, password: str) -> str:
+        """Test the connection and return the device serial number.
+
+        Raises a ConfigFlowError on failure.
+        """
+        controller = AsyncRainbirdController(
+            AsyncRainbirdClient(
+                async_get_clientsession(self.hass),
+                host,
+                password,
+            )
+        )
+        try:
+            async with async_timeout.timeout(TIMEOUT_SECONDS):
+                return await controller.get_serial_number()
+        except asyncio.TimeoutError as err:
+            raise ConfigFlowError(
+                f"Timeout connecting to Rain Bird controller: {str(err)}",
+                "timeout_connect",
+            ) from err
+        except RainbirdApiException as err:
+            raise ConfigFlowError(
+                f"Error connecting to Rain Bird controller: {str(err)}",
+                "cannot_connect",
+            ) from err
+
+    async def async_step_import(self, config: dict[str, Any]) -> FlowResult:
+        """Import a config entry from configuration.yaml."""
+        for entry in self._async_current_entries():
```

Comment:
Can we use this helper?

https://github.com/home-assistant/core/blob/bdcccd9d830d45d965d3da1a907fdd89d991eac8/homeassistant/config_entries.py#L1376-L1380